server/executor/transaction_runner.go

```diff
@@ -121,93 +117,81 @@ func (r persistentTransactionRunner) runTransaction(ctx context.Context, run mod
 		func(m subscription.Message) error {
-			updatedRun := m.Content.(model.Run)
+			testRun := m.Content.(model.Run)
+			if testRun.LastError != nil {
+				tr.State = model.TransactionRunStateFailed
+				tr.LastError = testRun.LastError
+			}
 
-			for i, run := range transactionRun.StepRuns {
-				if run.ID == updatedRun.ID && run.TestID == updatedRun.TestID {
-					transactionRun.StepRuns[i] = createStepRun(updatedRun)
-					break
-				}
+			tr, err = r.updateStepRun(ctx, tr, step, testRun)
+			if err != nil {
+				done <- true
+				return err
 			}
 
 			r.subscriptionManager.PublishUpdate(subscription.Message{
-				ResourceID: transactionRun.ResourceID(),
+				ResourceID: tr.ResourceID(),
 				Type:       "result_update",
-				Content:    transactionRun,
+				Content:    tr,
 			})
 
-			return nil
-		}),
-	)
-
-	err = r.updater.Update(ctx, transactionRun)
-	if err != nil {
-		return model.TransactionRun{}, fmt.Errorf("could not update transaction run: %w", err)
-	}
-
-	var wg sync.WaitGroup
-	wgPtr := &wg
-	wg.Add(1)
-
-	r.subscriptionManager.Subscribe(testRun.TransactionStepResourceID(), subscription.NewSubscriberFunction(
-		func(m subscription.Message) error {
-			defer wgPtr.Done()
-			runResult := m.Content.(RunResult)
+			fmt.Println("ACA", testRun.State, testRun.State.IsFinal())
+			if testRun.State.IsFinal() {
+				done <- true
+			}
 
-			return r.updateTransactionRun(ctx, stepIndex, &transactionRun, runResult)
+			return nil
 		}),
 	)
+	// TODO: this will block indefinitely. we need to set a timeout or something
+	<-done
 
-	wg.Wait()
-
-	return transactionRun, nil
+	return tr, err
 }
 
-func createStepRun(testRun model.Run) model.TransactionStepRun {
-	return model.TransactionStepRun{
-		ID:          testRun.ID,
-		TestID:      testRun.TestID,
-		Result:      *testRun.Results,
-		State:       testRun.State,
-		Environment: testRun.Environment,
-		Outputs:     testRun.Outputs,
+func (r persistentTransactionRunner) updateStepRun(ctx context.Context, tr model.TransactionRun, step int, run model.Run) (model.TransactionRun, error) {
+	if len(tr.Steps) <= step {
+		tr.Steps = append(tr.Steps, model.Run{})
 	}
-}
 
-func (r persistentTransactionRunner) updateTransactionRun(ctx context.Context, stepIndex int, transactionRun *model.TransactionRun, runResult RunResult) error {
-	testRun, err := runResult.Run, runResult.Err
+	tr.Steps[step] = run
+	err := r.updater.Update(ctx, tr)
 	if err != nil {
-		transactionRun.State = model.TransactionRunStateFailed
-		transactionRun.StepRuns[stepIndex] = createStepRun(testRun)
-		r.updater.Update(ctx, *transactionRun)
-
-		return err
+		return model.TransactionRun{}, fmt.Errorf("could not update transaction run: %w", err)
 	}
 
-	stepRun := createStepRun(testRun)
+	return tr, nil
+}
 
-	transactionRun.StepRuns[stepIndex] = stepRun
+// func (run TransactionRun) InjectOutputsIntoEnvironment(env Environment) Environment {
```

Comment:
Where are you injecting the outputs into the environment?

Reply:
uppps, fixed!